core/src/main/java/org/apache/calcite/rel/rules/ExpandDisjuctionRule.java

```diff
@@ -0,0 +1,379 @@
+import org.apache.calcite.rel.type.RelDataTypeField;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexTableInputRef;
+import org.apache.calcite.rex.RexUtil;
+import org.apache.calcite.tools.RelBuilder;
+import org.apache.calcite.util.ControlFlowException;
+import org.apache.calcite.util.ImmutableBitSet;
+
+import com.google.common.collect.Lists;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.immutables.value.Value;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Rule to expand disjuction in condition of a {@link Filter} or {@link Join},
+ * It makes sense to make this optimization part of the predicate pushdown. For example:
+ *
+ * <blockquote><pre>{@code
+ * select t1.name from t1, t2
+ * where t1.id = t2.id
+ * and (
+ *  (t1.id > 20 and t2.height < 50)
+ *  or
+ *  (t1.weight < 200 and t2.sales > 100)
+ * )
+ * }</pre></blockquote>
+ *
+ * <p>we can expand to obtain the condition
+ *
+ * <blockquote><pre>{@code
+ * t1.id > 20 or t1.weight < 200
+ * t2.height < 50 or t2.sales > 100
+ * }</pre></blockquote>
+ *
+ * <p>new generated predicates are redundant, but they could be pushed down to
+ * scan operator of t1/t2 and reduce the cardinality.
+ *
+ * <p>This rule should only be applied once to avoid generate same redundant expression and
+ * it should be used before {@link CoreRules#FILTER_INTO_JOIN} and {@link CoreRules#JOIN_CONDITION_PUSH}.
+ *
+ * @see CoreRules#EXPAND_FILTER_DISJUCTION
+ * @see CoreRules#EXPAND_JOIN_DISJUCTION
+ */
+@Value.Enclosing
+public class ExpandDisjuctionRule
+    extends RelRule<ExpandDisjuctionRule.Config>
+    implements TransformationRule {
+
+  /**
+   * Creates a ExpandDisjuctionRule.
+   */
+  protected ExpandDisjuctionRule(Config config) {
+    super(config);
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    config.matchHandler().accept(this, call);
+  }
+
+  /**
+   * Expand predicates for condition of a {@link Filter} or {@link Join}.
+   *
+   * @param condition   Condition to be expanded in Filter or Join
+   * @param relNode     The Filter or Join node
+   * @param fieldList   The field list of the Filter or Join inputs. Build the referenced columns
+   *                    in the predicate as RexInputRef and  find out which table the RexInputRef
+   *                    comes from via {@link RelMetadataQuery#getExpressionLineage}
+   * @param relBuilder  Builder
+   * @param mq          RelMetadataQuery, which is used to get the expression lineage
+   *
+   * @return Expanded predicates
+   */
+  private RexNode apply(
+      RexNode condition,
+      RelNode relNode,
+      List<RelDataTypeField> fieldList,
+      RelBuilder relBuilder,
+      RelMetadataQuery mq) {
+    HashMap<Integer, RexTableInputRef.RelTableRef> inputRefToTableRef = new HashMap<>();
```

Comment:
Prefer interface to concrete classes unless there is a good reason not to, you might consider using the `final` modifier to stress the fact that the variable is not going to be overwritten later on (the method is long enough that such hint helps), in this and other cases where relevant.